tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+Tests_SRS_INTERLOCKED_43_040: [ interlocked_compare_exchange_16 shall return the initial value of *destination. ]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_16_not_equal_case)
+{
+    ///arrange
+    volatile int16_t destination = INT16_MAX;
+    int16_t comperand = INT16_MIN;
+    int16_t exchange = INT16_MIN;
+
+    ///act
+    int16_t return_val = interlocked_compare_exchange_16(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX, destination);
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_008 [interlocked_compare_exchange_64 shall compare the 64-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_041: [ interlocked_compare_exchange_64 shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_64_equal_case)
+{
+    ///arrange
+    volatile int64_t destination = INT64_MAX;
+    int64_t comperand = INT64_MAX;
+    int64_t exchange = INT64_MIN;
+
+    ///act
+    int64_t return_val = interlocked_compare_exchange_64(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MIN, destination);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, return_val);
+}
+/*
+Tests_SRS_INTERLOCKED_43_008 [interlocked_compare_exchange_64 shall compare the 64-bit integers pointed to by destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_041: [ interlocked_compare_exchange_64 shall return the initial value of *destination. ]
+*/
+TEST_FUNCTION(interlocked_compare_exchange_64_not_equal_case)
+{
+    ///arrange
+    volatile int64_t destination = INT64_MAX;
+    int64_t comperand = INT64_MIN;
+    int64_t exchange = INT64_MIN;
+
+    ///act
+    int64_t return_val = interlocked_compare_exchange_64(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, destination);
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_010 [interlocked_compare_exchange_pointer shall compare the pointers destination and comperand. If they are equal, *destination is set to exchange. These operations are performed atomically.]
+Tests_SRS_INTERLOCKED_43_042: [ interlocked_compare_exchange_pointer shall return the initial value of *destination. ]
+*/
+
+TEST_FUNCTION(interlocked_compare_exchange_pointer_equal_case)
+{
+    ///arrange
+    int value1 = 1;
+    int value2 = 2;
+    volatile void* destination = &value1;
+    void* comperand = &value1;
+    void* exchange = &value2;
+
+    ///act
+    void* return_val = interlocked_compare_exchange_pointer(&destination, exchange, comperand);
```

Comment:
Interesting how the compile decided not to barf here.

As I read destination, it is a pointer to something volatile. Itself, the pointer, is not volatile.

However, &destination is pointer to a pointer to something volatile. But it should be a pointer to a volatile pointer, (exactly because it is changed by the function call).

Interesting... 

Would you mind taking a look?

Reply:
seems like a case for here and everywhere with  pointers.... 

---
In reply to: [427736331](https://github.com/Azure/azure-c-util/pull/43#discussion_r427736331) [](ancestors = 427736331)

Reply:
Yeah, Parth, this is the source of your other build failure I believe. I wrote out a few more examples in a quick console application, which may be helpful to understand which type is correct:

[https://gist.github.com/mattdurak/3f54cde77767185d3f1070c6edad7b59](https://gist.github.com/mattdurak/3f54cde77767185d3f1070c6edad7b59)


---
In reply to: [427736421](https://github.com/Azure/azure-c-util/pull/43#discussion_r427736421) [](ancestors = 427736421,427736331)